### PR TITLE
feat: add MCP header validation hardening

### LINKED ISSUE #149: MCP Protocol Support
- [ ] Use Kuadrant MCP Gateway if feasible
- [ ] Model Context Protocol proxy and routing
- [ ] JSON-RPC 2.0 message parsing and forwarding
- [ ] MCP-specific filter hooks (tool invocation, context management)
- [ ] Integration tests + example config

### PR DESCRIPTION
**Pull request by nerdalert**
This is the second PR for the MCP classifier filter to enforce stricter validation (was supposed to be closer to half but the first PR scope creeped):

- Reject `tools/call`, `prompts/get`, and `resources/read` when the required `params.name` or `params.uri` field is missing from the body (respects `on_invalid: continue` for pass-through pipelines)
- Reject spurious `Mcp-Name` headers on methods that have no body-derived name (e.g. `tools/list` with `Mcp-Name: evil`) under default `mismatch: reject`

  Refs praxis-proxy/ai#149

### CODE REVIEW

File: filter/src/builtins/http/payload_processing/mcp/tests.rs
Comment (twghu) [inline filter/src/builtins/http/payload_processing/mcp/tests.rs:1]: - missing message on line 708 compare for example line 626

File: filter/src/builtins/http/payload_processing/mcp/mod.rs
Comment (twghu) [inline filter/src/builtins/http/payload_processing/mcp/mod.rs:1]: Line 228, should the comment be skip requests with methods .. ?

File: tests/integration/tests/suite/mcp.rs
Comment (twghu) [inline tests/integration/tests/suite/mcp.rs:1]: Is this enough tests?

**Review by twghu — CHANGES_REQUESTED** [review]
just some  questions,  see file comments.

File: filter/src/builtins/http/payload_processing/mcp/mod.rs [hunk 1]
```diff
@@ -221,6 +225,19 @@ fn handle_non_mcp(config: &McpConfig) -> Result<FilterAction, FilterError> {
     }
 }
 
+/// Reject or skip methods that require a name/URI when extraction failed.
+fn reject_missing_required_selector(mcp: &McpEnvelope, config: &McpConfig) -> Option<FilterAction> {
+    let requires = mcp.method.requires_name() || mcp.method.requires_uri();
+    if requires && mcp.name.is_none() {
+        match config.on_invalid {
+            InvalidMcpBehavior::Reject => Some(FilterAction::Reject(Rejection::status(400))),
```
Comment (shaneutt) [inline filter/src/builtins/http/payload_processing/mcp/mod.rs]: On rejection we return a flat 400, no body, but sibling functions like `mcp_header_mismatch_rejection` returns a structured JSON-RPC error body. We need to consistently provide errors, here we should probably return a JSON-RPC `InvalidParams`.

File: filter/src/builtins/http/payload_processing/mcp/tests.rs [hunk 2]
```diff
@@ -607,6 +607,107 @@ async fn synthesize_injects_standard_mcp_headers() {
     );
 }
 
+// -----------------------------------------------------------------------------
+// Required Name/URI Tests
+// -----------------------------------------------------------------------------
+
```
Comment (shaneutt) [inline filter/src/builtins/http/payload_processing/mcp/tests.rs:613]: Need a test for `tools/call` with _no params key at all_ as opposed to empty params.

Also need a negative test for non-string name values.

File: filter/src/builtins/http/payload_processing/mcp/tests.rs [hunk 3]
```diff
@@ -607,6 +607,107 @@ async fn synthesize_injects_standard_mcp_headers() {
...
+async fn spurious_mcp_name_ignored_when_mismatch_ignore() {
+    let filter = make_filter(r#"{"header_validation": {"mismatch": "ignore"}}"#);
+    let body_str = r#"{"jsonrpc":"2.0","id":1,"method":"tools/list"}"#;
+    let req = make_mcp_request(&[("mcp-name", "evil")]);
+    let mut ctx = crate::test_utils::make_filter_context(&req);
+    let mut body = Some(Bytes::from(body_str));
+
+    let action = filter.on_request_body(&mut ctx, &mut body, true).await.unwrap();
+
+    assert!(matches!(action, FilterAction::Release));
```
Comment (shaneutt) [inline filter/src/builtins/http/payload_processing/mcp/tests.rs]: Project conventions ask for assertion messages, this would be a good place to add one.

File: tests/integration/tests/suite/mcp.rs [hunk 4]
```diff
@@ -225,6 +225,54 @@ fn mcp_standard_headers_preserved_internal_stripped() {
...
+    assert_eq!(
+        parse_status(&raw),
+        400,
+        "spurious Mcp-Name on nameless method should be rejected in default mode"
+    );
+}
+
 // -----------------------------------------------------------------------------
 // Fragmented / Oversized Body Tests
 // -----------------------------------------------------------------------------
```
Comment (shaneutt) [inline tests/integration/tests/suite/mcp.rs:343]: We could use more integration tests here that show that the params are getting to the backend properly, with a mock MCP server, etc.

**Review by shaneutt — CHANGES_REQUESTED** [review]

File: filter/src/builtins/http/payload_processing/mcp/tests.rs
Comment (nerdalert) [reply filter/src/builtins/http/payload_processing/mcp/tests.rs:1]: Done ty. Added the missing assertion message on the mismatch-ignore test.

File: filter/src/builtins/http/payload_processing/mcp/mod.rs
Comment (nerdalert) [reply filter/src/builtins/http/payload_processing/mcp/mod.rs:1]: Reworded the selector comment to explain why selector-bearing methods cannot be trusted when the selector is absent or malformed.

File: tests/integration/tests/suite/mcp.rs
Comment (nerdalert) [reply tests/integration/tests/suite/mcp.rs:1]: Done ty. Expanded the integration coverage with backend-forwarding proof plus missing/no params and non-string name rejection cases.

File: filter/src/builtins/http/payload_processing/mcp/mod.rs (hunk 1 above)
Comment (nerdalert) [reply filter/src/builtins/http/payload_processing/mcp/mod.rs]: Added Missing required MCP selectors now return a InvalidParams error body with the original request ID, instead of a flat 400.

File: filter/src/builtins/http/payload_processing/mcp/tests.rs (hunk 2 above)
Comment (nerdalert) [reply filter/src/builtins/http/payload_processing/mcp/tests.rs:613]: Added coverage for tools/call with no params key and for non-string params.name, the non-string case also exercises string ID escaping.

File: filter/src/builtins/http/payload_processing/mcp/tests.rs (hunk 3 above)
Comment (nerdalert) [reply filter/src/builtins/http/payload_processing/mcp/tests.rs]: Added, ty.

File: tests/integration/tests/suite/mcp.rs (hunk 4 above)
Comment (nerdalert) [reply tests/integration/tests/suite/mcp.rs:343]: Added an integration test using the echo backend to prove valid MCP params are forwarded upstream unchanged, plus integration coverage for invalid missing/non-string params.
